File: p2/p252/CanAttendMeetings.java
package leetcode.p2.p252;

import java.util.Arrays;
import java.util.Comparator;
import java.util.PriorityQueue;

public class CanAttendMeetings {
    /**
     * Feb 27, 2022 01:49
     * Priority Queue
     * Time Complexity: O(nlogn)
     * Space Complexity: O(n)
     */
    public static boolean canAttendMeetings1(int[][] intervals) {
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(o -> o[0]));
        pq.addAll(Arrays.asList(intervals));
        int[] prev = pq.poll();
        while (!pq.isEmpty()) {
            int[] temp = pq.poll();
            if (temp[0] < prev[1]) {
                return false;
            }
            prev = temp;
        }
        return true;
    }

    /**
     * Feb 27, 2022 01:53
     * Arrays.sort()
     * Time Complexity: O(nlogn)
     * Space Complexity: O(1)
     */

    public static boolean canAttendMeetings2(int[][] intervals) {
        Arrays.sort(intervals, Comparator.comparingInt(o -> o[0]));
        for (int i = 0; i < intervals.length - 1; i++) {
            if (intervals[i][1] > intervals[i + 1][0]) {
                return false;
            }
        }
        return true;
    }


    public static void main(String[] args) {

    }

}

File: p2/p252/can_attend_meetings.py
import heapq
from typing import List


class Solution:
    """
    Mar 02, 2025 22:06
    Time Complexity: O(nlogn)
    Space Complexity: O(n)
    """

    def canAttendMeetings(self, intervals: List[List[int]]) -> bool:
        sorted_intervals = sorted(intervals, key=lambda x: x[1])
        for i in range(1, len(sorted_intervals)):
            if sorted_intervals[i][0] < sorted_intervals[i - 1][1]:
                return False

        return True


class Solution1:
    """
    Mar 02, 2025 22:17
    Time Complexity: O(nlogn)
    Space Complexity: O(n)
    """

    def canAttendMeetings(self, intervals: List[List[int]]) -> bool:
        if len(intervals) == 0:
            return True
        heap = [(x[1], x) for x in intervals]
        heapq.heapify(heap)

        prev = heapq.heappop(heap)[1]
        while heap:
            curr = heapq.heappop(heap)[1]
            if prev[1] > curr[0]:
                return False

            prev = curr

        return True


if __name__ == '__main__':
    obj = Solution1()
    print(obj.canAttendMeetings([[7,10],[2,4]]))